### PR TITLE
Contextualize rework

### PR DESCRIPTION
**Pull request by smaldini**
Draft: https://gist.github.com/smaldini/d407cc9f9a1d13bbde6dcd93166df8c7
WIP : 
- [x] add log() support : log propagate and and parent events
- [x] add Flux/Mono contextMap
- [x] Streamable Context ?
- [x] polish API (public Contextualized ? factorize implemented interfaces ?... )
- [x] verify merge/fork use cases (processor, publish, merge, zip...)
- [x] add `Hooks.onSubscriber( (s, c) -> s)`
- [x] review (packages, api, feedback with #337 #117 #210 )

### CODE REVIEW

**Comment by OlegDokuka** [comment]
What is the use case of Context?

**Comment by smaldini** [comment]
@OlegDokuka we'll refine it this month but it is an ad hoc attribute/context immutable transport upstream and downstream (push updates) and can be propagated to inner sources like in concatMap, zip or flatMap. It is meant to be the alternative to thread-local in a reactive environment that is missed by some spring and beyond, but also for our own ecosystem (netty, kafka) or for overriding hooks (local drop error consumer, local queue factory, local delay error strategy).

This combined to the new Hooks.onSubscriber matching the inner or outer end subscriber allows to add adhoc serviceability for zipkin for instance. Or pass Principal or Transaction.

**Comment by OlegDokuka** [comment]
Got it, thanks!

File: reactor-core/src/main/java/reactor/core/publisher/MonoOperator.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.util.Objects;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Publisher;
+import reactor.core.Scannable;
+
+/**
+ * A decorating {@link Mono} {@link Publisher} that exposes {@link Mono} API over an
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/MonoOperator.java]: rewrite to reference `MonoOperator` rather than `MonoSource` ;)

File: reactor-core/src/main/java/reactor/util/context/ContextN.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	ContextN(Object key1, Object value1, Object key2, Object value2) {
+		super(2, 1f);
+		super.put(key1, value1);
+		super.put(key2, value2);
+	}
+
+	ContextN(Map<Object, Object> map, Object key, @Nullable Object value) {
+		super(map.size() + 1, 1f);
+		putAll(map);
+		super.put(key, value);
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/util/context/ContextN.java]: if `value` is null, should it be put in the map?

File: reactor-core/src/main/java/reactor/core/publisher/FluxSink.java [hunk 3]
```diff
@@ -31,6 +32,16 @@
 public interface FluxSink<T> {
 
 	/**
+	 * Immediately propagate a {@link Context} to the child {@link Subscriber} given an
+	 * eventually non empty parent {@link Context}.
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxSink.java]: replace "an eventually" by "a potentially"

File: reactor-core/src/main/java/reactor/core/publisher/FluxWindowPredicate.java [hunk 4]
```diff
@@ -288,7 +295,7 @@ public Object scanUnsafe(Attr key) {
 
 		@Override
 		public Stream<? extends Scannable> inners() {
-			return window == null ? Stream.empty() : Stream.of(window);
+			return Stream.of(window);
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxWindowPredicate.java]: 🐞 the "rebase" commit somehow changed that line. couldn't find a parallel PR where this changed, and I think it is a definite mistake (`window` can definitely be `null`). Change back to `return window == null ? Stream.empty() : Stream.of(window)`

File: reactor-core/src/main/java/reactor/core/publisher/Hooks.java [hunk 5]
```diff
@@ -99,6 +101,27 @@ public static void onOperatorError(BiFunction<? super Throwable, Object, ?
 	}
 
 	/**
+	 * Set a global "assembly" hook to intercept signals produced by the passed
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Hooks.java]: (nitpick: weird line breaks all over this javadoc and method)

File: reactor-core/src/main/java/reactor/core/publisher/Hooks.java [hunk 6]
```diff
@@ -130,6 +153,16 @@ public static void resetOnOperator() {
 	}
 
 	/**
+	 * Reset global "assembly" hook tracking
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Hooks.java]: copy pasted javadoc

File: reactor-core/src/main/java/reactor/core/publisher/Mono.java [hunk 7]
```diff
@@ -3345,26 +3387,21 @@ public String toString() {
 			@Nullable Consumer<? super T> onNext,
 			@Nullable Consumer<? super Throwable> onError,
 			@Nullable Runnable onComplete,
-			@Nullable Runnable onAfterTerminate,
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Mono.java]: I would have maybe opened a separate issue for that orthogonal change, :+1: for the change itself

File: reactor-core/src/main/java/reactor/core/publisher/MonoPeek.java [hunk 8]
```diff
@@ -102,7 +99,7 @@ public Runnable onCompleteCall() {
 	@Override
 	@Nullable
 	public Runnable onAfterTerminateCall() {
-		return onAfterTerminateCall;
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/MonoPeek.java]: maybe mark this method as "unused" in the javadoc to remove any ambiguity (`FluxPeek` uses it, `MonoPeek` doesn't and favors `MonoPeekTerminal` instead)

File: reactor-core/src/main/java/reactor/core/publisher/MonoPeekFuseable.java [hunk 9]
```diff
@@ -105,7 +102,7 @@ public Runnable onCompleteCall() {
 	@Override
 	@Nullable
 	public Runnable onAfterTerminateCall() {
-		return onAfterTerminateCall;
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/MonoPeekFuseable.java]: maybe mark this method as "unused" in the javadoc to remove any ambiguity (`FluxPeekFuseable` uses it, `MonoPeekFuseable` doesn't and favors `MonoPeekTerminal` instead)

File: reactor-core/src/main/java/reactor/core/publisher/MonoSink.java [hunk 10]
```diff
@@ -38,6 +40,16 @@
 	void success();
 
 	/**
+	 * Immediately propagate a {@link Context} to the child {@link Subscriber} given an
+	 * eventually non empty parent {@link Context}.
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/MonoSink.java]: "an eventually" => "a potentially"

File: reactor-core/src/main/java/reactor/core/publisher/MonoSource.java [hunk 11]
```diff
@@ -22,37 +22,27 @@
...
  * @param <I> delegate {@link Publisher} type
  * @param <O> produced type
  */
+@Deprecated
 public class MonoSource<I, O> extends Mono<O> implements Scannable {
 
 	protected final Publisher<? extends I> source;
 
 	/**
-	 * Unchecked wrap of {@link Publisher} as {@link Mono}, supporting {@link Fuseable} sources
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/MonoSource.java]: deprecate rather than remove? (public API)

File: reactor-core/src/main/java/reactor/core/publisher/SignalLogger.java [hunk 12]
```diff
@@ -191,8 +200,21 @@ else if (level == Level.SEVERE) {
 		}
 		return null;
 	}
+@Override
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/SignalLogger.java]: `\n\t` + `@Nullable` :)

File: reactor-core/src/main/java/reactor/core/publisher/SignalLogger.java [hunk 13]
```diff
@@ -191,8 +200,21 @@ else if (level == Level.SEVERE) {
...
+	public Consumer<? super Context> onContextPropagateCall() {
+		if ((options & CONTEXT_PROPAGATE) == CONTEXT_PROPAGATE && (level != Level.INFO || log.isInfoEnabled())) {
+			return c -> log(SignalType.ON_CONTEXT, c, source);
+		}
+		return null;
+	}
 
-	String subscriptionAsString(@Nullable Subscription s) {
+	@Override
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/SignalLogger.java]: `@Nullable`

File: reactor-core/src/main/java/reactor/core/publisher/SignalLogger.java [hunk 14]
```diff
@@ -191,8 +200,21 @@ else if (level == Level.SEVERE) {
...
+	public Consumer<? super Context> onContextParentCall() {
+		if ((options & CONTEXT_PARENT) == CONTEXT_PARENT && (level != Level.INFO || log.isInfoEnabled())) {
+			return c -> log(SignalType.CURRENT_CONTEXT, c, source);
+		}
+		return null;
+	}	String subscriptionAsString(@Nullable Subscription s) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/SignalLogger.java]: missing newline x2 (note a parallel PR #659 changes `subscriptionAsString` to static)

File: reactor-core/src/main/java/reactor/core/publisher/SignalPeek.java [hunk 15]
```diff
@@ -97,4 +98,19 @@
 		return null;
 	}
 
+	/**
+	 * A task that will run on {@link Context} propagation from upstream to downstream
+	 * @return A task that will run on {@link Context} propagation from upstream to downstream
+	 */
+	default Consumer<? super Context> onContextPropagateCall(){
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/SignalPeek.java]: `@Nullable`

File: reactor-core/src/main/java/reactor/core/publisher/SignalPeek.java [hunk 16]
```diff
@@ -97,4 +98,19 @@
...
+		return null;
+	}
+
+	/**
+	 * A task that will run on {@link Context} read from downstream to upstream
+	 * @return A task that will run on {@link Context} propagation from upstream to downstream
+	 */
+	default Consumer<? super Context> onContextParentCall(){
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/SignalPeek.java]: `@Nullable`

File: reactor-core/src/main/java/reactor/core/publisher/SignalType.java [hunk 17]
```diff
@@ -51,9 +51,17 @@
 	 */
 	ON_COMPLETE,
 	/**
-	 * A signal that can be propagated AFTER an operator has terminated
+	 * A signal when an operator completes
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/SignalType.java]: why remove the `AFTER` part? it is significant.

File: reactor-core/src/main/java/reactor/core/publisher/SignalType.java [hunk 18]
```diff
@@ -51,9 +51,17 @@
...
+	 */
+	AFTER_TERMINATE,
+	/**
+	 * A context read signal that is pulled
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/SignalType.java]: rephrase to "A signal that context was read / pulled from parent"?

File: reactor-core/src/main/java/reactor/core/publisher/SignalType.java [hunk 19]
```diff
@@ -51,9 +51,17 @@
...
+	 */
+	CURRENT_CONTEXT,
+	/**
+	 * A context signal that can be propagated from parent source
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/SignalType.java]: rephrase to "A signal that a context was propagated from parent to source"? (or is it from source to parent?)

File: reactor-core/src/main/java/reactor/core/publisher/SignalType.java [hunk 20]
```diff
@@ -70,6 +78,10 @@ public String toString() {
 				return "request";
 			case CANCEL:
 				return "cancel";
+			case CURRENT_CONTEXT:
+				return "parent";
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/SignalType.java]: `"parent"` is a bit vague... why not eg. `"currentContext"`?

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java [hunk 21]
```diff
@@ -3333,7 +3374,7 @@ public final T blockLast(Duration timeout) {
 	 * the same key (yet element keys can repeat in the overall sequence)
 	 */
 	public final <V> Flux<T> distinctUntilChanged(Function<? super T, ? extends V> keySelector) {
-		return distinctUntilChanged(keySelector, Objects::equals);
+		return distinctUntilChanged(keySelector, equalPredicate());
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Flux.java]: this equals function change is orthogonal... shelve it and isolate it in another PR / commit on master

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java [hunk 22]
```diff
@@ -7526,6 +7571,11 @@ public boolean test(O o) {
 	}
 
 	@SuppressWarnings("unchecked")
+	static <U, V> BiPredicate<U, V> equalPredicate() {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Flux.java]: this equals function change is orthogonal... shelve it and isolate it in another PR / commit on master

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java [hunk 23]
```diff
@@ -7544,6 +7594,7 @@ public boolean test(O o) {
 	static final BooleanSupplier ALWAYS_BOOLEAN_SUPPLIER = () -> true;
 	@SuppressWarnings("rawtypes")
 	static final Function        HASHCODE_EXTRACTOR      = Object::hashCode;
+	static final BiPredicate     OBJECT_EQUAL            = Object::equals;
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Flux.java]: this equals function change is orthogonal... shelve it and isolate it in another PR / commit on master

File: reactor-core/src/main/java/reactor/core/publisher/Mono.java [hunk 24]
```diff
@@ -2116,7 +2150,7 @@ public T block(Duration timeout) {
 	 * otherwise
 	 */
 	public final Mono<Boolean> hasElement() {
-		return onAssembly(new MonoHasElements<>(this));
+		return onAssembly(new MonoHasElement<>(this));
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Mono.java]: this MonoHasElement change is orthogonal... shelve it and isolate it in another PR / commit on master

File: reactor-core/src/main/java/reactor/core/publisher/MonoHasElement.java [hunk 25]
```diff
@@ -0,0 +1,84 @@
+/*
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/MonoHasElement.java]: this MonoHasElement change is orthogonal... shelve it and isolate it in another PR / commit on master
